From 2256b4671d48bcddff5145904d50c91fd91f00a2 Mon Sep 17 00:00:00 2001
From: Niko Tyni <ntyni@debian.org>
Date: Fri, 11 Feb 2022 21:55:29 +0200
Subject: [PATCH] syswrite test: unclog the pipe harder

Reading out 4 kB from a full pipe is not enough to unblock
it on all systems.

On Linux, the pipe size is by default 16 * page size, and
at least the Debian (and apparently Alpine too) ppc64le port
has a page size of 64 kB instead of the more usual 4 kB.

Bug: https://rt.cpan.org/Public/Bug/Display.html?id=140489
Bug-Debian: https://bugs.debian.org/1004146
---
 lib/Test/Future/IO/Impl.pm | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- a/lib/Test/Future/IO/Impl.pm
+++ b/lib/Test/Future/IO/Impl.pm
@@ -307,15 +307,18 @@
       pipe my ( $rd, $wr ) or die "Cannot pipe() - $!";
       $wr->blocking( 0 );
 
+      my $i = 0;
       # Attempt to fill the pipe
-      $wr->syswrite( "X" x 4096 ) for 1..256;
+      $i++ while $wr->syswrite( "X" x 4096 );
+      $! == Errno::EAGAIN or $! == Errno::EWOULDBLOCK or
+        die "Expected EAGAIN, got $!";
 
       my $f = Future::IO->syswrite( $wr, "more" );
 
       ok( !$f->is_ready, '$f is still pending' );
 
       # Now make some space
-      $rd->read( my $buf, 4096 );
+      $rd->read( my $buf, 4096 ) while !$f->is_ready and $i-- > 0;
 
       is( scalar $f->get, 4, 'Future::IO->syswrite yields written count' );
    }
